File: cat-client/src/main/java/com/dianping/cat/message/internal/AbstractMessage.java
/*
 * Copyright (c) 2011-2018, Meituan Dianping. All Rights Reserved.
 *
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements. See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License. You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.dianping.cat.message.internal;

import java.nio.charset.Charset;

import io.netty.buffer.ByteBuf;
import io.netty.buffer.ByteBufAllocator;

import com.dianping.cat.message.Message;
import com.dianping.cat.message.spi.codec.PlainTextMessageCodec;

public abstract class AbstractMessage implements Message {
	protected String m_status = "unset";

	private String m_type;

	private String m_name;

	private long m_timestampInMillis;

	private CharSequence m_data;

	private boolean m_completed;

	public AbstractMessage(String type, String name) {
		m_type = String.valueOf(type);
		m_name = String.valueOf(name);
		m_timestampInMillis = MilliSecondTimer.currentTimeMillis();
	}

	@Override
	public void addData(String keyValuePairs) {
		if (m_data == null) {
			m_data = keyValuePairs;
		} else if (m_data instanceof StringBuilder) {
			((StringBuilder) m_data).append('&').append(keyValuePairs);
		} else {
			StringBuilder sb = new StringBuilder(m_data.length() + keyValuePairs.length() + 16);

			sb.append(m_data).append('&');
			sb.append(keyValuePairs);
			m_data = sb;
		}
	}

	@Override
	public void addData(String key, Object value) {
		if (m_data instanceof StringBuilder) {
			((StringBuilder) m_data).append('&').append(key).append('=').append(value);
		} else {
			String str = String.valueOf(value);
			int old = m_data == null ? 0 : m_data.length();
			StringBuilder sb = new StringBuilder(old + key.length() + str.length() + 16);

			if (m_data != null) {
				sb.append(m_data).append('&');
			}

			sb.append(key).append('=').append(str);
			m_data = sb;
		}
	}

	@Override
	public CharSequence getData() {
		if (m_data == null) {
			return "";
		} else {
			return m_data;
		}
	}

	public void setData(String str) {
		m_data = str;
	}

	@Override
	public String getName() {
		return m_name;
	}

	public void setName(String name) {
		m_name = name;
	}

	@Override
	public String getStatus() {
		return m_status;
	}

	@Override
	public void setStatus(Throwable e) {
		m_status = e.getClass().getName();
	}

	@Override
	public long getTimestamp() {
		return m_timestampInMillis;
	}

	@Override
	public void setTimestamp(long timestamp) {
		m_timestampInMillis = timestamp;
	}

	@Override
	public String getType() {
		return m_type;
	}

	public void setType(String type) {
		m_type = type;
	}

	@Override
	public boolean isCompleted() {
		return m_completed;
	}

	public void setCompleted(boolean completed) {
		m_completed = completed;
	}

	@Override
	public boolean isSuccess() {
		return Message.SUCCESS.equals(m_status);
	}

	@Override
	public void setStatus(String status) {
		m_status = status;
	}

	@Override
	public String toString() {
		PlainTextMessageCodec codec = new PlainTextMessageCodec();
		ByteBuf buf = ByteBufAllocator.DEFAULT.buffer();

		codec.encodeMessage(this, buf);
		codec.reset();
		return buf.toString(Charset.forName("utf-8"));
	}

	@Override
	public void setSuccessStatus() {
		m_status = SUCCESS;
	}

}

File: lib/csharp/src/CatClient/Message/Internals/AbstractMessage.cs
using System;
using System.Text;
using Org.Unidal.Cat.Message.Spi.Codec;
using Org.Unidal.Cat.Util;

namespace Org.Unidal.Cat.Message.Internals
{
    public abstract class AbstractMessage : IMessage
    {
        private String _mName;
        private String _mType;
        private bool _mCompleted;
        private StringBuilder _mData;

        private String _mStatus = "unset";

        protected AbstractMessage(String type, String name)
        {
            _mType = type;
            _mName = name;
            TimestampInMicros = MilliSecondTimer.UnixNowMicroSeconds();
        }

        /// <summary>
        ///   其实是Ticks除以10
        /// </summary>
        protected long TimestampInMicros { get; private set; }

        #region IMessage Members

        public String Data
        {
            get { return _mData == null ? "" : _mData.ToString(); }
        }

        public String Name
        {
            get { return _mName; }
            set { _mName = value ?? CatConstants.NULL_STRING; }
        }

        public String Status
        {
            get { return _mStatus; }

            set
            {
                _mStatus = (value == null) ? "unset" : (value.Length > 64 ? value.Substring(0, 64) : value);
            }
        }

        /// <summary>
        ///   其实是Ticks除以10000
        /// </summary>
        public long Timestamp
        {
            get { return TimestampInMicros/1000L; }
            set { TimestampInMicros = value*1000L; }
        }

        public String Type
        {
            get { return _mType; }
            set { _mType = value ?? CatConstants.NULL_STRING; }
        }

        public virtual void AddData(String keyValuePairs)
        {
            try
            {
                // No meaning to add null as data.
                if (null == keyValuePairs)
                    return;

                if (_mData == null)
                {
                    _mData = new StringBuilder(keyValuePairs);
                }
                else
                {
                    _mData.Append(keyValuePairs);
                }
            }
            catch (Exception ex)
            {
                Cat.lastException = ex;
            }
        }

        public virtual void AddData(String key, Object value)
        {
            try
            {
                if (_mData == null)
                {
                    _mData = new StringBuilder();
                }
                else if (_mData.Length > 0)
                {
                    _mData.Append('&');
                }

                _mData.Append(key).Append('=').Append(value);
            }
            catch (Exception ex)
            {
                Cat.lastException = ex;
            }
        }

        public virtual void Complete()
        {
            SetCompleted(true);
        }

        public bool IsCompleted()
        {
            return _mCompleted;
        }

        public bool IsSuccess()
        {
            return CatConstants.SUCCESS.Equals(_mStatus);
        }

        public void SetStatus(Exception e)
        {
            if (null != e)
            {
                _mStatus = e.GetType().FullName;
            }
        }

        #endregion

        public void SetCompleted(bool completed)
        {
            _mCompleted = completed;
        }

        public override String ToString()
        {
            try
            {
                PlainTextMessageCodec codec = new PlainTextMessageCodec();
                ChannelBuffer buf = new ChannelBuffer(8192);

                codec.EncodeMessage(this, buf);
                buf.Reset();

                return buf.ToString();
            }
            catch (Exception ex)
            {
                Cat.lastException = ex;
                return "";
            }
        }

        public int EstimateByteSize()
        {
            return (Type == null ? 0 : Type.Length)
                + (Name == null ? 0 : Name.Length)
                + (Data == null ? 0 : Data.Length);
                //+ (_mStatus == null ? 0 : _mStatus.Length);
        }
    }
}